### ISSUE TITLE
Reuse containers

### ISSUE THREAD

**Issue by tisonkun** [testcontainers/testcontainers-rs#742]
Like https://java.testcontainers.org/features/reuse/

I currently use `bollard` directly to reuse containers between tests, while noticed that `testcontainers` encapsulation is valuable and ergonomic, like all the code to support `get_host_port_ipv4`.

So if `testcontainers` can support reuse containers, users like me can benefit from both features.

**Comment by tisonkun** [comment]
My current code in gist - https://gist.github.com/tisonkun/c2ac3154bd6bea003c6097b195a9e32d

**Comment by DDtKey** [comment]
Hi 👋 
Thank you for raising the issue!

Yes, it makes sense to support and should be simple

**Comment by gurinderu** [comment]
How do you want to make a reuse ID? Should it be calculated automatically from the source code & call site meta or provided by hand?

**Comment by gurinderu** [comment]
Also, it is not clear to understand when to stop the container, lets say we run nextest with the same reuse id in a serial way.  Nextest use process-per-test strategy and we don't have ryuk

**Comment by DDtKey** [comment]
Reusable containers are not expected to be cleaned. 

Documentation from dotnet impl:
> Enabling reuse will disable the resource reaper, meaning the resource will not be cleaned up

and also from Java:
> Reusable containers are not suited for CI usage and as an experimental feature not all Testcontainers features are fully working (e.g., resource cleanup or networking).

You can check [java](https://java.testcontainers.org/features/reuse/) and [dotnet](https://dotnet.testcontainers.org/api/resource_reuse/) docs, because it describes how it works.

**Comment by tisonkun** [comment]
> Reusable containers are not suited for CI usage

Just a comment to this statement: We reuse containers in the CI environment, and the runner will destroy everything once the job is finished. So it works well for our CI usage :D

**Comment by DDtKey** [comment]
Yes, actually it should work if your runners are properly configured and don't store the state between different jobs. E.g DnD approach will work

**Comment by the-wondersmith** [comment]
@DDtKey I've actually been working on implementing support for this. Would you accept a PR for the feature?

**Comment by DDtKey** [comment]
@the-wondersmith I will be happy to review the PR.

Feel free to open 🙂

**Comment by the-wondersmith** [comment]
> @the-wondersmith I will be happy to review the PR.
> 
> Feel free to open 🙂

Will do, as soon as I'm able 😁

**Comment by DDtKey** [comment]
Actually I'd like to see how exactly the labels are going to be used, before merging this part. For now it's more like black box without justification.

I mean this PR has references(in documentation) to some internal usage which does not exist yet and it's harder to review without full context

Also I can provide some guidance or highlight something while having the whole context

We still can merge them as 2 separate PRs if agreed, but you can open second one as soon as you are ready.

Thank you!

**Comment by the-wondersmith** [comment]
@DDtKey Apologies for the delay (life and all 😅). The [follow up PR](https://github.com/testcontainers/testcontainers-rs/pull/757) for a working implementation of reusable containers is up. 

So far as I'm able to tell, I've maintained behavior parity with the implementations in other languages. Originally, I'd included global reaping behavior as well that used [`ctor`](https://crates.io/crates/ctor) to clean up containers at the end of a test *session*. I stripped that out though, as I kinda realized it muddied things up too much for a solid single-feature PR review.

Anyway, eager for your thoughts, whenever you have time 😁

**Comment by DDtKey** [comment]
@the-wondersmith thank you! I'll review as soon as I can.

There is no a goal to clean reusable containers, at least for now. Because it's pretty complicated, some users may want them to be alive for a long time and reuse once per hour for example.

Also, what I thinking about is to allow controlling this behavior via ENV variable, it's common case that I don't want to change code, but run them in reusable mode only on my machine.

**Comment by the-wondersmith** [comment]
> [@the-wondersmith](https://github.com/the-wondersmith) thank you! I'll review as soon as I can.
> 
> There is no a goal to clean reusable containers, at least for now. Because it's pretty complicated, some users may want them to be alive for a long time and reuse once per hour for example.
> 
> Also, what I thinking about is to allow controlling this behavior via ENV variable, it's common case that I don't want to change code, but run them in reusable mode only on my machine.

Yeah, that's kinda what I figured as well. I hadn't considered that that part though. Adding env var based control shouldn't be hard at all - just pull the value from env instead of hard-coding `false` where it's set by default 🤔. 

I'll add it shortly, if you like 😁

**Comment by DDtKey** [comment]
Yes, that totally make sense to implement ENV variable control separately in follow up PR

**Comment by DDtKey** [comment]
I'll review the PRs this week, sorry for the delay 🙏

**Comment by the-wondersmith** [comment]
@tisonkun reusable containers was [officially merged](https://github.com/testcontainers/testcontainers-rs/pull/757) this morning 😁. If you can, please give it a whirl and report back with any thoughts?

For the sake of transparency, there is still planned follow-up work on the feature (e.g. env var control). Still though, feedback would be great 🙂

**Comment by tisonkun** [comment]
@the-wondersmith Cool! Thanks everyone helps in this thread.

I'll test this feature in this week.

**Comment by the-wondersmith** [comment]
@DDtKey where did we end up on the remaining / follow-up work? i.e. what's left to do, what can I (potentially) start on?

**Comment by DDtKey** [comment]
Thanks one more time @the-wondersmith for implementing this feature! 🙏 

I think follow-ups were mentioned here: 

https://github.com/testcontainers/testcontainers-rs/pull/757#pullrequestreview-2504927584

**Comment by the-wondersmith** [comment]
> Thanks one more time [@the-wondersmith](https://github.com/the-wondersmith) for implementing this feature! 🙏

It was absolutely my pleasure 😁

> I think follow-ups were mentioned here:
> 
> [#757 (review)](https://github.com/testcontainers/testcontainers-rs/pull/757#pullrequestreview-2504927584)
> 
> > * env-variable control
> > * proper reaper for "current session" (ryuk)

Where's the best place to discuss these? Should we continue here, or maybe open a separate tracking issue?

**Comment by DDtKey** [comment]
I think it's better to create (follow-up/sub)-issues and discuss them separately

**Comment by the-wondersmith** [comment]
> I think it's better to create (follow-up/sub)-issues and discuss them separately
> 
> Because basic implementation has been provided already

Agreed. Spun up an issue [here](https://github.com/testcontainers/testcontainers-rs/issues/761). On re-read though, maybe I should have made 2? Look at the new one and let me know - I'll spin the `ryuk` discussion to its own issue if that'd be easier / better.

**Comment by tisonkun** [comment]
I tried at https://github.com/tisonkun/morax/pull/19

@the-wondersmith it seems not work. See the PR for reproduce and fail logs.

**Comment by tisonkun** [comment]
Also, there is a lack of documents about `ReuseDirective`. I don't know the different between `CurrentSession` and `Always` clearly.

**Comment by the-wondersmith** [comment]
@tisonkun thank you for this feedback! That's a good callout, by the by, and I guess we should clarify in the docs.

The *intended* behavior is *supposed* to be:

- `ReuseDirective::Never`: previous standard behavior, i.e. just start a new container every time
- `ReuseDirective::Always`:
  - try find the first running container whose configuration matches that of the `ContainerRequest` being started
  - if none are found, start a new container
  - if exactly one is found, use it
  - if *more* than one is found, use the container that was most recently created
- `ReuseDirective::CurrentSession`: same rules as `Always`, but with the added restriction that the container must have been created by the current "session" (run of a test / test suite / application / etc ...)
  - *generally* this added restriction means that no running containers will match the filters upon first inspection, due to the [session id](https://github.com/testcontainers/testcontainers-rs/blob/main/testcontainers/src/runners/async_runner.rs#L39) being globally unique for each "run", meaning a container will pretty much always be started for every initial (unique) request. Subsequent requests should then reuse the created container.

I've looked over your test logs (thank you *so* much for including those by the way), and I think I figured out what the issue is. I'll push up a branch in a bit and shoot you over some instructions for how to test it, if you're up for that?

**Comment by tisonkun** [comment]
> if you're up for that?

Sure. I'm glad to test it out.

**Comment by the-wondersmith** [comment]
@tisonkun Apologies for the delay in getting back to this. 

Could you update your [`Cargo.toml`](https://github.com/tisonkun/morax/blob/try-testconatiners-reuse/Cargo.toml#L83) to point to the branch I've got up with what I *think* is the fix here? Should look like this:

```toml
testcontainers = { git = "https://github.com/the-wondersmith/testcontainers-rs.git", branch = "fix/reusable-containers", features = [
  "blocking",
  "reusable-containers",
] }
```

I've pulled down your `morax` repo locally and made the update, and while the test suite *is* still failing, it seems to be failing differently (i.e. for **_non_**-`testcontainers` related reasons). I'm just not familiar enough with whatever `morax` is or is doing to be able to tell 😅.

If it passes muster for you, I'll push up a PR to `testcontainers` itself and get the fix(es) upstreamed.

**Comment by tisonkun** [comment]
@the-wondersmith I tried in https://github.com/tisonkun/morax/pull/23.

It seems the test can run but the container is removed after test passed. So that we don't reuse the container between test cases.

**Comment by tisonkun** [comment]
@the-wondersmith would you have time to take a look at the feedback above?

**Comment by the-wondersmith** [comment]
> [@the-wondersmith](https://github.com/the-wondersmith) would you have time to take a look at the feedback above?

@tisonkun Yes, sorry. I've been out sick this week 🤒. I'll make time for it ASAP and get a follow up posted

**Comment by tisonkun** [comment]
@the-wondersmith Take care of yourself! It's not a hurry; just a kindly reminder :P

**Comment by DDtKey** [comment]
@the-wondersmith take care and get well soon!
Let me know if you need help, I can take on the task if you are unable to

**Comment by cbrevik** [comment]
Wrt. `ReuseDirective::Always` it is noted above here that one tries to find the first **running** container whose configuration matches that of the `ContainerRequest` being started. 

Would it be possible to support reuse of stopped containers as well? The use-case here is that I've used TestContainers to spin up a database for local development, and I'd like to reuse the same container after reboots so I do not lose data.

**Comment by DDtKey** [comment]
Hi @cbrevik 👋 

> Would it be possible to support reuse of stopped containers as well? The use-case here is that I've used TestContainers to spin up a database for local development, and I'd like to reuse the same container after reboots so I do not lose data

Based on the use case described it looks like you can achieve that without reusable containers.

You can utilize named docker volumes for your postgres container. It's supported pretty well

**Comment by cbrevik** [comment]
> Hi [@cbrevik](https://github.com/cbrevik) 👋
> 
> > Would it be possible to support reuse of stopped containers as well? The use-case here is that I've used TestContainers to spin up a database for local development, and I'd like to reuse the same container after reboots so I do not lose data
> 
> Based on the use case described it looks like you can achieve that without reusable containers.
> 
> You can utilize named docker volumes for your postgres container. It's supported pretty well

Hi @DDtKey, thanks! I haven't tested that out for this specific use case, so I'll be sure to do so. 

Just to note though, I've followed the pattern of using reusable containers for this sort of use case with `testcontainers-node` and `testcontainers-dotnet`, which supports reusing stopped containers. Not sure about the Java or other languages.

**Comment by cbrevik** [comment]
Unsure if this belongs as an issue here, or in the community crates-repo, but using named mounts with the Postgres-package works the first time I start the container with a volume. But on the second time (in other words, the volume exists already), the container readiness check eventually times out. 

Which is kind of weird, since checking the logs from the Docker-container itself, and interacting with it, it seems like it is ready. It prints the required log line that the Psql-testcontainer checks for as well, `database system is ready to accept connections`.

I've  made a repro here: https://github.com/cbrevik/named_mount/blob/main/src/main.rs

**Comment by the-wondersmith** [comment]
@DDtKey apologies for having let this slip - life uhhh... happens. I'm back though, and am tackling this. I (apparently) deleted the branch with a/the fix on it 🤦, so I've got to dig back in and refresh myself.

> (@cbrevik) Just to note though, I've followed the pattern of using reusable containers for this sort of use case with `testcontainers-node` and `testcontainers-dotnet`, which supports reusing stopped containers. Not sure about the Java or other languages.

IIRC, the plan was always to bring everything into alignment *eventually*, with the existing implementation serving as a foundation upon which to iterate. tl;dr I don't want to speak for @DDtKey and/or the `testcontainers-rs` maintainers, but I'd imagine that if the reuse behavior in `testcontainers-node` and `testcontainers-dotnet` works some particular way, `testcontainers-rs` will work its way towards the same.

**Comment by DDtKey** [comment]
Just to clarify - support of stopped containers makes sense to me.
I just wanted to point to the alternative solution for that particular usecase to unblock the user

**Comment by the-wondersmith** [comment]
@DDtKey Ah ha! Thank goodness for GitHub's [`Activity` feature](https://github.com/orgs/community/discussions/55884#discussioncomment-9331142) - I was able to un-delete [the branch](https://github.com/the-wondersmith/testcontainers-rs/tree/fix/reusable-containers) I referenced [here](https://github.com/testcontainers/testcontainers-rs/issues/742#issuecomment-2599856689).

Looking at the diff, I remember what I figured out was the issue - filtering by `name` wouldn't be correct in a `testcontainers` context, we *actually* need to filter by [`ancestor`](https://docs.docker.com/reference/cli/docker/container/ls/#filter:~:text=%2C%20or%20dead-,ancestor,-Filters%20containers%20which) using the `ContainerRequest`'s image and tag.

Now I just need to dig into the post-test container deletion @tisonkun described 😅

**Comment by the-wondersmith** [comment]
@tisonkun would it be possible to set up a working block with you to look at your [`morax` branch](https://github.com/tisonkun/morax/pull/23) and get this working as expected?

**Comment by tisonkun** [comment]
@the-wondersmith I can check it out in the next week. Is the new version released or there is a testcontainers-rs branch or commit I can use to test?

**Comment by the-wondersmith** [comment]
@tisonkun

> `{ ... }` Is the new version released or there is a testcontainers-rs branch or commit I can use to test?

It would be the same fork+branch patch [I referenced originally](
https://github.com/testcontainers/testcontainers-rs/issues/742#issuecomment-2599856689).

Full disclosure: I *have* made a couple updates there, but the fundamental reuse behavior hasn't changed from your last test run (I don't think). Really, I'm just trying to get a handle on the difference between the expected and observed behavior in a real-world usage setting (like what you've got in `morax`), so I can 1) *document* the expected behavior, and 2) update the tests to match and subsequently *enforce*  the correct behavior moving forward. 

One thing that *might* help shed some light is if you make sure to set the `RUST_LOG` env var to `info,testcontainers=trace,testcontainers_modules=trace` and then run the test with `-- --nocapture`. You *might* need to add something like [`test-log`](https://crates.io/crates/test-log) to the tests though, depending on how `morax` handles logging and such. Anyway, if you do, you should see some trace logs  from the `Drop` impl that'll illuminate what's going on in the cleanup logic.

**Comment by tisonkun** [comment]
@the-wondersmith 

Seems containers can be reused if run with `cargo test`. But I need to run each test per process for some reasons and thus use `cargo nextest`, then it outputs as below and cannot reuse containers between tests:

```
────────────
 Nextest run ID 04c3283b-ace4-41cf-a8ec-3ab66b661637 with nextest profile: default
    Starting 7 tests across 14 binaries
       START             behavior-tests::simple_pubsub test_simple_pubsub

running 1 test
2025-10-27T22:30:31.704750+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:103 Registering IPv4 port mapping: 5432/tcp -> 32795
2025-10-27T22:30:31.778379+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:111 Registering IPv6 port mapping: 5432/tcp -> 32795
2025-10-27T22:30:31.778428+08:00[Asia/Shanghai] DEBUG testcontainers::core::containers::async_container: async_container.rs:385 Waiting for container 3b77642c3aa9ca40169b3d87ed9f47d08f48e9b305fe196235d27e35023fb7b8 to be ready
2025-10-27T22:30:34.688111+08:00[Asia/Shanghai] DEBUG testcontainers::core::logs: logs.rs:100 Message found 1 times after comparing 8 lines
2025-10-27T22:30:34.688681+08:00[Asia/Shanghai] DEBUG testcontainers::core::containers::async_container: async_container.rs:394 Container 3b77642c3aa9ca40169b3d87ed9f47d08f48e9b305fe196235d27e35023fb7b8 is now ready!
2025-10-27T22:30:34.693420+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:103 Registering IPv4 port mapping: 5432/tcp -> 32795
2025-10-27T22:30:34.693495+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:111 Registering IPv6 port mapping: 5432/tcp -> 32795
2025-10-27T22:30:34.878206+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:103 Registering IPv4 port mapping: 9000/tcp -> 32796
2025-10-27T22:30:34.878291+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:111 Registering IPv6 port mapping: 9000/tcp -> 32796
2025-10-27T22:30:34.878355+08:00[Asia/Shanghai] DEBUG testcontainers::core::containers::async_container: async_container.rs:385 Waiting for container 83e7ed2f0faa7bb3b48ed3d3082f5527dfda1566540ff09785a6720bfe58e931 to be ready
2025-10-27T22:30:35.077645+08:00[Asia/Shanghai] DEBUG testcontainers::core::logs: logs.rs:100 Message found 1 times after comparing 8 lines
2025-10-27T22:30:35.077735+08:00[Asia/Shanghai] DEBUG testcontainers::core::containers::async_container: async_container.rs:394 Container 83e7ed2f0faa7bb3b48ed3d3082f5527dfda1566540ff09785a6720bfe58e931 is now ready!
2025-10-27T22:30:35.081157+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:103 Registering IPv4 port mapping: 9000/tcp -> 32796
2025-10-27T22:30:35.081214+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:111 Registering IPv6 port mapping: 9000/tcp -> 32796
2025-10-27T22:30:35.083780+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:103 Registering IPv4 port mapping: 5432/tcp -> 32795
2025-10-27T22:30:35.083841+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:111 Registering IPv6 port mapping: 5432/tcp -> 32795
2025-10-27T22:30:35.086528+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:103 Registering IPv4 port mapping: 9000/tcp -> 32796
2025-10-27T22:30:35.086580+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:111 Registering IPv6 port mapping: 9000/tcp -> 32796
2025-10-27T22:30:35.121421+08:00[Asia/Shanghai]  INFO morax_runtime [redacted]
2025-10-27T22:30:35.128070+08:00[Asia/Shanghai]  INFO morax_meta [redacted]
2025-10-27T22:30:35.178927+08:00[Asia/Shanghai]  INFO morax_meta [redacted]
2025-10-27T22:30:35.307986+08:00[Asia/Shanghai]  INFO morax_meta [redacted]
2025-10-27T22:30:35.308027+08:00[Asia/Shanghai]  INFO morax_meta [redacted]
2025-10-27T22:30:35.392976+08:00[Asia/Shanghai]  INFO morax_runtime [redacted]
2025-10-27T22:30:35.393023+08:00[Asia/Shanghai]  INFO morax_runtime [redacted]
2025-10-27T22:30:35.393480+08:00[Asia/Shanghai]  INFO morax_runtime [redacted]
2025-10-27T22:30:35.393948+08:00[Asia/Shanghai]  INFO poem::server: server.rs:182 listening addr=socket://192.168.1.3:64974
2025-10-27T22:30:35.393998+08:00[Asia/Shanghai]  INFO poem::server: server.rs:184 server started
2025-10-27T22:30:35.394017+08:00[Asia/Shanghai]  INFO morax_server [redacted]
2025-10-27T22:30:35.394463+08:00[Asia/Shanghai]  INFO morax_server [redacted]
2025-10-27T22:30:35.394491+08:00[Asia/Shanghai]  INFO poem::server: server.rs:191 initiate graceful shutdown timeout_in_seconds=30.0
2025-10-27T22:30:35.394601+08:00[Asia/Shanghai]  INFO poem::server: server.rs:268 server stopped
2025-10-27T22:30:35.394718+08:00[Asia/Shanghai]  INFO morax_server [redacted]
2025-10-27T22:30:35.504738+08:00[Asia/Shanghai] DEBUG testcontainers::core::containers::async_container: async_container.rs:323 Deleting docker container 83e7ed2f0faa7bb3b48ed3d3082f5527dfda1566540ff09785a6720bfe58e931
2025-10-27T22:30:35.637133+08:00[Asia/Shanghai] TRACE testcontainers::core::containers::async_container: async_container.rs:429 Declining to reap container 83e7ed2f0faa7bb3b48ed3d3082f5527dfda1566540ff09785a6720bfe58e931 (marked as reuse = Always)
2025-10-27T22:30:35.637218+08:00[Asia/Shanghai] DEBUG testcontainers::core::containers::async_container: async_container.rs:323 Deleting docker container 3b77642c3aa9ca40169b3d87ed9f47d08f48e9b305fe196235d27e35023fb7b8
2025-10-27T22:30:35.790826+08:00[Asia/Shanghai] TRACE testcontainers::core::containers::async_container: async_container.rs:429 Declining to reap container 3b77642c3aa9ca40169b3d87ed9f47d08f48e9b305fe196235d27e35023fb7b8 (marked as reuse = Always)
test test_simple_pubsub ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out; finished in 4.35s

        PASS [   4.370s] behavior-tests::simple_pubsub test_simple_pubsub
       START             behavior-tests::simple_pubsub test_simple_pubsub_dup

running 1 test
2025-10-27T22:30:35.957388+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:103 Registering IPv4 port mapping: 5432/tcp -> 32797
2025-10-27T22:30:35.972147+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:111 Registering IPv6 port mapping: 5432/tcp -> 32797
2025-10-27T22:30:35.972188+08:00[Asia/Shanghai] DEBUG testcontainers::core::containers::async_container: async_container.rs:385 Waiting for container 8263347bc03f85f8e0027e020ab2154033f9608e132154b36e2f09258eaa987e to be ready
2025-10-27T22:30:38.723569+08:00[Asia/Shanghai] DEBUG testcontainers::core::logs: logs.rs:100 Message found 1 times after comparing 8 lines
2025-10-27T22:30:38.723729+08:00[Asia/Shanghai] DEBUG testcontainers::core::containers::async_container: async_container.rs:394 Container 8263347bc03f85f8e0027e020ab2154033f9608e132154b36e2f09258eaa987e is now ready!
2025-10-27T22:30:38.727090+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:103 Registering IPv4 port mapping: 5432/tcp -> 32797
2025-10-27T22:30:38.727170+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:111 Registering IPv6 port mapping: 5432/tcp -> 32797
2025-10-27T22:30:38.914423+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:103 Registering IPv4 port mapping: 9000/tcp -> 32798
2025-10-27T22:30:38.914495+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:111 Registering IPv6 port mapping: 9000/tcp -> 32798
2025-10-27T22:30:38.914555+08:00[Asia/Shanghai] DEBUG testcontainers::core::containers::async_container: async_container.rs:385 Waiting for container c50a56f21126aa3ebcba712f684df3228f4572bcef795689ce2902e869179b37 to be ready
2025-10-27T22:30:39.017177+08:00[Asia/Shanghai] DEBUG testcontainers::core::logs: logs.rs:100 Message found 1 times after comparing 8 lines
2025-10-27T22:30:39.017277+08:00[Asia/Shanghai] DEBUG testcontainers::core::containers::async_container: async_container.rs:394 Container c50a56f21126aa3ebcba712f684df3228f4572bcef795689ce2902e869179b37 is now ready!
2025-10-27T22:30:39.019755+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:103 Registering IPv4 port mapping: 9000/tcp -> 32798
2025-10-27T22:30:39.019819+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:111 Registering IPv6 port mapping: 9000/tcp -> 32798
2025-10-27T22:30:39.021372+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:103 Registering IPv4 port mapping: 5432/tcp -> 32797
2025-10-27T22:30:39.021407+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:111 Registering IPv6 port mapping: 5432/tcp -> 32797
2025-10-27T22:30:39.022872+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:103 Registering IPv4 port mapping: 9000/tcp -> 32798
2025-10-27T22:30:39.022907+08:00[Asia/Shanghai] DEBUG testcontainers::core::ports: ports.rs:111 Registering IPv6 port mapping: 9000/tcp -> 32798
2025-10-27T22:30:39.066214+08:00[Asia/Shanghai]  INFO morax_runtime [redacted]
2025-10-27T22:30:39.070691+08:00[Asia/Shanghai]  INFO morax_meta [redacted]
2025-10-27T22:30:39.114829+08:00[Asia/Shanghai]  INFO morax_meta [redacted]
2025-10-27T22:30:39.226456+08:00[Asia/Shanghai]  INFO morax_meta [redacted]
2025-10-27T22:30:39.226493+08:00[Asia/Shanghai]  INFO morax_meta [redacted]
2025-10-27T22:30:39.288164+08:00[Asia/Shanghai]  INFO morax_runtime [redacted]
2025-10-27T22:30:39.288198+08:00[Asia/Shanghai]  INFO morax_runtime [redacted]
2025-10-27T22:30:39.288531+08:00[Asia/Shanghai]  INFO morax_runtime [redacted]
2025-10-27T22:30:39.288865+08:00[Asia/Shanghai]  INFO poem::server: server.rs:182 listening addr=socket://192.168.1.3:64992
2025-10-27T22:30:39.288901+08:00[Asia/Shanghai]  INFO poem::server: server.rs:184 server started
2025-10-27T22:30:39.288918+08:00[Asia/Shanghai]  INFO morax_server [redacted]
2025-10-27T22:30:39.289289+08:00[Asia/Shanghai]  INFO morax_server [redacted]
2025-10-27T22:30:39.289306+08:00[Asia/Shanghai]  INFO poem::server: server.rs:191 initiate graceful shutdown timeout_in_seconds=30.0
2025-10-27T22:30:39.289395+08:00[Asia/Shanghai]  INFO poem::server: server.rs:268 server stopped
2025-10-27T22:30:39.289463+08:00[Asia/Shanghai]  INFO morax_server [redacted]
2025-10-27T22:30:39.368353+08:00[Asia/Shanghai] DEBUG testcontainers::core::containers::async_container: async_container.rs:323 Deleting docker container c50a56f21126aa3ebcba712f684df3228f4572bcef795689ce2902e869179b37
2025-10-27T22:30:39.485403+08:00[Asia/Shanghai] TRACE testcontainers::core::containers::async_container: async_container.rs:429 Declining to reap container c50a56f21126aa3ebcba712f684df3228f4572bcef795689ce2902e869179b37 (marked as reuse = Always)
2025-10-27T22:30:39.485478+08:00[Asia/Shanghai] DEBUG testcontainers::core::containers::async_container: async_container.rs:323 Deleting docker container 8263347bc03f85f8e0027e020ab2154033f9608e132154b36e2f09258eaa987e
2025-10-27T22:30:39.631811+08:00[Asia/Shanghai] TRACE testcontainers::core::containers::async_container: async_container.rs:429 Declining to reap container 8263347bc03f85f8e0027e020ab2154033f9608e132154b36e2f09258eaa987e (marked as reuse = Always)
test test_simple_pubsub_dup ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out; finished in 3.82s

        PASS [   3.834s] behavior-tests::simple_pubsub test_simple_pubsub_dup
       START             morax-runtime global::tests::test_spawn_block_on

running 1 test
test global::tests::test_spawn_block_on ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 4 filtered out; finished in 0.01s

        PASS [   0.024s] morax-runtime global::tests::test_spawn_block_on
       START             morax-runtime global::tests::test_spawn_from_blocking

running 1 test
test global::tests::test_spawn_from_blocking ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 4 filtered out; finished in 0.00s

        PASS [   0.009s] morax-runtime global::tests::test_spawn_from_blocking
       START             morax-runtime runtime::tests::test_block_on

running 1 test
test runtime::tests::test_block_on ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 4 filtered out; finished in 0.05s

        PASS [   0.060s] morax-runtime runtime::tests::test_block_on
       START             morax-runtime runtime::tests::test_spawn_blocking

running 1 test
test runtime::tests::test_spawn_blocking ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 4 filtered out; finished in 0.00s

        PASS [   0.010s] morax-runtime runtime::tests::test_spawn_blocking
       START             morax-runtime runtime::tests::test_spawn_join

running 1 test
test runtime::tests::test_spawn_join ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 4 filtered out; finished in 0.00s

        PASS [   0.006s] morax-runtime runtime::tests::test_spawn_join
────────────
     Summary [   8.317s] 7 tests run: 7 passed, 0 skipped
```

Code: https://github.com/tisonkun/morax/pull/27

**Comment by the-wondersmith** [comment]
@tisonkun

> Seems containers can be reused if run with `cargo test`. But I need to run each test per process for some reasons and thus use `cargo nextest`, then it outputs as below and cannot reuse containers between tests `{ ... }`

<img width="500px" src="https://c.tenor.com/551UXO_M05cAAAAC/tenor.gif" />

[`carg-nextest`](https://nexte.st/) is *100%* where it's **_at_**. In fairness, I haven't tried any of the `reusable-containers` stuff with it though.

Okie doke, the context and test logs at least give me a starting point to debug from. I'll dig in as soon as I'm able and see what I can figure out.

**Comment by tisonkun** [comment]
Today I have some time to figure it out.

Here is a fix - https://github.com/testcontainers/testcontainers-rs/pull/887
